### DISCUSSION TITLE
Utility voltage fluctuating wildly

### DISCUSSION THREAD

**Discussion by bantaj** [jgyates/genmon#1150] [Questions]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).

### Version

Version V1.19.03

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have submitted logs

### What is/are your question(s)?

I have 2 separate GenMon instances monitoring 2 separate Generac 14Kw generators.  One of them is showing very large fluctuations in the utility voltage monitor (typically ranging from 200-260v).  The following graph is from Home Assistant where the data is captured via MQTT and logged.
<img width="571" alt="image" src="https://github.com/user-attachments/assets/cbb7efc3-0abe-416d-a7bd-44f0e718cac3">

The other instance of GenMon on a different generator shows utility voltage very stable.

The main difference in setup between the 2 is the generator showing the fluctuations is using an esp32 as a serial bridge to transmit the data from the Generac to the RasPi but I dont think that would cause this.

Using a voltmeter on a 240 receptacle does not show this same behavior with the readings steady at ~240.

Any suggestions for further troubleshooting?

**Comment by jgyates** [comment]
Does the communications stats on the Monitor page show any timeout errors in the Communication Stats section? If the wifi signal is not optimal, this can cause invalid data in some cases. The know issues wiki page has more info, see issues "16 communication errors"

https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues#16-communication-errors


If this is what is happening, the short story on this is that if the wifi signal is weak the 802.11 wifi protocol utilizes retries (or resending of the data) to compensate for a bad signal and this can cause the window of retries to overlap with the modbus timeout window. This coupled with the fact that the modbus protocol does not support canceling a transaction, this can create a situation where the data from one modbus transaction get mixed with the previous modbus request. The wiki has some mitigation that you could try.

Improving your wifi signal, or moving to wired Ethernet or serial would also solve the issue.

**Reply by bantaj** [reply]
I am not seeing any timeout errors on the Monitor page and wifi signal strength (for the rs232 bridge) has been consistent between -50 to -55 dBm.

If it were a communication error, would other data show similar fluctuations (battery voltage for instance)?  

Communication Stats :
        Packet Count : M: 26165, S: 26164
        CRC Errors : 0
        CRC Percent Errors : 0.00%
        Timeout Errors : 0
        Timeout Percent Errors : 0.00%
        Modbus Exceptions : 0
        Validation Errors : 0
        Sync Errors : 0
        Invalid Data : 0
        Discarded Bytes : 0
        Comm Restarts : 0
        Packets Per Second : 16.19
        Average Transaction Time : 0.1206 sec
        Modbus Transport : TCP

**Reply by jgyates** [reply]
that is good news about no comm errors. we can rule that out. Would it be possible to swap the two genmon instances and see if the behavior follows the pi or the controller? 

It is possible that the voltage sensor on one of the controllers faulty, but this test would allow you to isolate the problem to either genmon/pi or the controller/generator. The utility voltage is just a value read via modbus and displayed so if there are no comm errors my guess is that this is likely the value the controller is returning. Are the two controllers running the same version of the firmware?
